### PR TITLE
Implement SIA-R70

### LINKED ISSUE #826: Implement SIA-R70
https://github.com/Siteimprove/sanshikan/blob/df450e27db11c4d7b36eb1d53fa0aecdb0f729ba/rules/sia-r70.md

Applicability:
* take inspiration from, e.g., R59 or R87 (both do the same thing in different ways…)

Expectation:
* Use `descendants()`, flat tree is obtained with options `flattened` and `nested`. See, e.g. Applicability of R2.
* use `isVisible` and `not(isIgnored)`
* Get inspiration from R1 (expectation) for grabbing elements by their name. `hasName` can take several arguments.

Extended diagnostic:
* We need to send the list of offending elements we've found.

### PR DESCRIPTION
**Pull request by Jym77**
Resolves #826 

#839 got botched with gazillion unrelated changes (likely messed up rebase from `develop` to `main`). Reaching the point where it's easier to light the copy/paste than to curse the rebase…

### CODE REVIEW

File: packages/alfa-rules/src/sia-r70/rule.ts [hunk 1]
```diff
@@ -0,0 +1,137 @@
+import { Rule, Diagnostic } from "@siteimprove/alfa-act";
+import { Document, Element, Namespace } from "@siteimprove/alfa-dom";
+import { Predicate } from "@siteimprove/alfa-predicate";
+import { Err, Ok } from "@siteimprove/alfa-result";
+import { Page } from "@siteimprove/alfa-web";
+import { List } from "@siteimprove/alfa-list";
```
Comment (kasperisager) [inline packages/alfa-rules/src/sia-r70/rule.ts:6]: This import can be removed.

File: packages/alfa-rules/src/sia-r70/rule.ts [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+    return (
+      value instanceof DeprecatedElements &&
+      value._message === this._message &&
+      Array.equals(value._errors, this._errors)
+    );
+  }
+
+  public *[Symbol.iterator](): Iterator<Element> {
+    yield* this._errors;
+  }
```
Comment (kasperisager) [inline packages/alfa-rules/src/sia-r70/rule.ts:118]:
```suggestion
  public *[Symbol.iterator](): Iterator<Element> {
    yield* this._errors;
  }

```

File: packages/alfa-rules/src/sia-r70/rule.ts [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+
+  private _errors: ReadonlyArray<Element>;
+  private constructor(message: string, errors: Array<Element>) {
+    super(message);
+    this._errors = errors;
+  }
+
+  public get errors(): ReadonlyArray<Element> {
+    return this._errors;
+  }
```
Comment (kasperisager) [inline packages/alfa-rules/src/sia-r70/rule.ts:105]:
```suggestion
  public get errors(): ReadonlyArray<Element> {
    return this._errors;
  }

```

File: packages/alfa-rules/src/sia-r70/rule.ts [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+  public equals(value: DeprecatedElements): boolean;
+  public equals(value: unknown): value is this;
+  public equals(value: unknown): boolean {
+    return (
+      value instanceof DeprecatedElements &&
+      value._message === this._message &&
+      Array.equals(value._errors, this._errors)
+    );
+  }
```
Comment (kasperisager) [inline packages/alfa-rules/src/sia-r70/rule.ts:114]:
```suggestion
  public equals(value: DeprecatedElements): boolean;

  public equals(value: unknown): value is this;

  public equals(value: unknown): boolean {
    return (
      value instanceof DeprecatedElements &&
      value._message === this._message &&
      Array.equals(value._errors, this._errors)
    );
  }
```

File: packages/alfa-rules/src/sia-r70/rule.ts [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+
+class DeprecatedElements extends Diagnostic implements Iterable<Element> {
+  public static of(
+    message: string,
+    errors: Iterable<Element> = []
+  ): DeprecatedElements {
+    return new DeprecatedElements(message, Array.from(errors));
+  }
+
+  private _errors: ReadonlyArray<Element>;
```
Comment (kasperisager) [inline packages/alfa-rules/src/sia-r70/rule.ts:97]:
```suggestion
  private _errors: ReadonlyArray<Element>;

```

File: packages/alfa-rules/src/sia-r70/rule.ts [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+
+export namespace Outcomes {
+  export const HasNoDeprecatedElement = Ok.of(
+    Diagnostic.of(`The document doesn't contain any deprecated element`)
+  );
+
+  export const HasDeprecatedElements = (errors: Iterable<Element>) =>
+    Err.of(
+      DeprecatedElements.of(
+        `The document contains deprecated elements: `,
```
Comment (kasperisager) [inline packages/alfa-rules/src/sia-r70/rule.ts:83]:
```suggestion
        `The document contains deprecated elements`,
```

File: packages/alfa-rules/src/sia-r70/rule.ts [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+          ),
+        };
+      },
+    };
+  },
+});
+
+export namespace Outcomes {
+  export const HasNoDeprecatedElement = Ok.of(
+    Diagnostic.of(`The document doesn't contain any deprecated element`)
```
Comment (kasperisager) [inline packages/alfa-rules/src/sia-r70/rule.ts:77]:
```suggestion
    Diagnostic.of(`The document doesn't contain any deprecated elements`)
```

**Review by kasperisager — CHANGES_REQUESTED** [review]
Just some minor nits. Looks great! 👌

File: packages/alfa-rules/src/sia-r70/rule.ts (hunk 6 above)
Comment (elenamongelli) [reply packages/alfa-rules/src/sia-r70/rule.ts:83]: We have that `errors` is an `Iterable<Element>`, so isn´t it more readable to leave : before the list is printed out? 
```
DeprecatedElements.of(
        `The document contains deprecated elements: `,
        errors
      )
```

Comment (kasperisager) [reply packages/alfa-rules/src/sia-r70/rule.ts:82]: The message is used independently of the diagnostic data and so the two won't necessarily be output together.

**Review by kasperisager — DISMISSED** [review]

**Review by kasperisager — APPROVED** [review]

**Pull request merged**